McvManagerBotModule: Add ability to build additional MCVs

LipkeGu opened the pull request:
This commit gives the bot the ability to build additional MCVs and send them out to random selected resource fields. Otherwise it deploys the MCV in/near the current base. The bot builds currently only 2 MCVs atm.

This PR picks "the Multible MCV logic" from #12186
superseeds #12186 and #17649

Testcase:
A Bot should send out MCVs nearby resource fields and deploy it (there).
Setting "MaxAllowedMCVs" to "1" will disable this additional functionality.

Steps:
[x] Add ability to build additional MCVs.
[X] Set the new deployed MCV as base center
[ ] Increase the building limit by the amount of Construction Yards

LipkeGu commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -63,7 +63,7 @@ public CPos GetRandomBaseCenter()
 		IBotRequestUnitProduction[] requestUnitProduction;
-
+		IResourceLayer resourceLayer;
 		CPos initialBaseCenter;
```
I would recommend to move `CPos initialBaseCenter` to player.cs as `CPos BaseCenter` and deduplicate the code in `BotMCVManagerModule` and `BotBasebuilderModule`. And also deduplicate `GetRandomBaseCenter()` as `GetBaseCenter(bool random = true)` to decide between `LastOrDefault()` and ` RandomorDefault() `

PunkPun replied:
bot code is separated from the rest of the codebase by design, I don't think we should move it to the player class

LipkeGu replied:
@phrohdoh have you an idea what we can do at this point?

abcdefg30 replied:
Why not keep track of the base centers inside the AI module instead of using the initial center from `Player.cs`?

LipkeGu commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -121,16 +121,20 @@ void IBotTick.BotTick(IBot bot)
+			return AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, owner);
+		}
+
+		bool AllowedToBuildMCV()
```
I renamed it to `AllowedToBuildMCV` because it handles also the check for `MaxNumberOfMCV`

LipkeGu commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -197,8 +201,24 @@ void DeployMcv(IBot bot, Actor mcv, bool move)
 				distanceToBaseIsImportant ? Info.MaxBaseRadius : world.Map.Grid.MaximumTileSearchRange);
+
+			// Otherwise select a location near resource fields.
+			if (resourceLayer != null && !distanceToBaseIsImportant)
```
It should be moved in near furture to a more usable and suitable Location/file. Only an Resourcefield is ugly... It would be better when the bot can decide between TargetLocations.

PunkPun replied:
it would be best to have some sort of clustering algorithm so that the AI would deploy to resource fields, and not the remains of a harvester

LipkeGu commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
This should be moved into a more base class. Something like "DeployableUnitManagerBotModule.cs" and use it also for other units like

* MobileWarfactory
* LimpedDrone
* MCVs
* and Others...

PunkPun commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -121,16 +121,20 @@ void IBotTick.BotTick(IBot bot)
-		bool ShouldBuildMCV()
+		int CountConstructionYards(Player owner)
+		{
+			return AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, owner);
```
shouldn't it count mcv's as well?

PunkPun commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -28,12 +28,12 @@ public class McvManagerBotModuleInfo : ConditionalTraitInfo
 
-		[Desc("Try to maintain at least this many ConstructionYardTypes, build an MCV if number is below this.")]
-		public readonly int MinimumConstructionYardCount = 1;
-
```
this will need an update rule

LipkeGu replied:
someone should explain it to me how to do this... :/

PunkPun replied:
your's should look similar to https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/UpdateRules/Rules/20231010/RemoveEditorSelectionLayerProperties.cs

the aforementioned class is reference in `UpdatePath.cs`

PunkPun commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -197,8 +201,24 @@ void DeployMcv(IBot bot, Actor mcv, bool move)
+				var nearbyResources = world.Map.FindTilesInAnnulus(selectedPosition.Value,
+					world.Map.Grid.MaximumTileSearchRange - 2, world.Map.Grid.MaximumTileSearchRange)
+					.Where(a => resourceLayer.GetResource(a).Type != null)
+					.Shuffle(world.LocalRandom).Take(6);
```
where's the handling for the case where no resources are found? Won't this crash?

LipkeGu committed:
McvManagerBotModule: Add ability to build more than one MCV

This commit gives the bot the ability to build additional MCVs and send them out to random selected resource fields.
Otherwise it deploys the MCV in/near the current base. This commit adapts the work of #17649

LipkeGu committed:
Add update rule

LipkeGu commented:
At this state the AI can build X MCVs and deploy them at the desiredlocation. To get this fully working the AI Basebuilding should be notified when an addition MCV has been deployed.

dnqbob commented:
Pal, you can click 
![image](https://github.com/OpenRA/OpenRA/assets/13763394/8991fe08-f9dd-4fd4-8ef4-08adb0c7611c)
To turn your PR as draft PR instead of closing and reopening.

LipkeGu commented:
The update rule is added btw...

PunkPun reviewed (commented):
I ran an AI game and not a single MCV was deployed by the AI. They all just sat in base

<img width="562" alt="MCV's" src="https://github.com/OpenRA/OpenRA/assets/37534529/9245284a-221e-480b-9e16-c7b5d11a86f2">

I also got this crash when one of the AI's got nuked
```
Exception of type `System.NullReferenceException`: Object reference not set to an instance of an object.
   at OpenRA.Mods.Common.Traits.McvManagerBotModule.GetBaseCenter(Boolean random) in /OpenRA/OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:line 66
   at OpenRA.Mods.Common.Traits.McvManagerBotModule.ChooseMcvDeployLocation(String actorType, CVec offset, Boolean distanceToBaseIsImportant) in /OpenRA/OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:line 220
   at OpenRA.Mods.Common.Traits.McvManagerBotModule.DeployMcv(IBot bot, Actor mcv, Boolean move) in /OpenRA/OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:line 166
```

ottelo9 commented on OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs:
```diff
@@ -195,10 +207,34 @@ void DeployMcv(IBot bot, Actor mcv, bool move)
+			var actors = world.FindActorsInCircle(wPos, newBaseRadius)
+				.Where(a => !a.Disposed);
+
+			var enemies = actors.Any(a => !a.Disposed && !player.IsAlliedWith(a.Owner) && a.Info.HasTraitInfo<BuildingInfo>());
```
var enemies = actors.Any(a => !a.Disposed && player.IsAlliedWith(a.Owner) &&

Remove the "!" before player.IsAlliedWith(a.Owner)

ottelo9 replied:
I got no crash if I nuke the AI.
